Added Admin Badge to Profile Popover

DSchalla opened the pull request:
#### Summary
This PR adds an admin badge to the profile popover show whether an user is an channel / team / system admin using a shield icon on the right side. The title of the icon shows the admin level of the user.

Preview:
![Example Image](https://i.imgur.com/9NPrc3q.png)

#### Ticket Link
N/A

#### Checklist
[Place an '[x]' (no spaces) in all applicable fields. Please remove unrelated fields.]
- [x] Ran `make check-style` to check for style errors (required for all pull requests)
- [x] Ran `make test` to ensure unit and component tests passed
- [x] Has UI changes

mattermod commented:
`Setup Test Server` label detected. Spinmint test server created if build succeeds (checks pass and no conflicts with base branch).

mattermod commented:
Spinmint test server created at: http://i-069b1a6304edde1b3.test.spinmint.com

Test Admin Account: Username: `sysadmin` | Password: `sysadmin`

Test User Account: Username: `user-1` | Password: `user-1`

Instance ID: i-069b1a6304edde1b3

mattermod commented:
Spinmint test server destroyed

mattermod commented:
`Setup Test Server` label detected. Spinmint test server created if build succeeds (checks pass and no conflicts with base branch).

mattermod commented:
Spinmint test server created at: http://i-0517fb71490894ab6.test.spinmint.com

Test Admin Account: Username: `sysadmin` | Password: `sysadmin`

Test User Account: Username: `user-1` | Password: `user-1`

Instance ID: i-0517fb71490894ab6

mattermod commented:
`Setup Test Server` label detected. Spinmint test server created if build succeeds (checks pass and no conflicts with base branch).

mattermod commented:
Spinmint test server created at: https://i-0705fa39969d0d176.test.spinmint.com

Test Admin Account: Username: `sysadmin` | Password: `sysadmin`

Test User Account: Username: `user-1` | Password: `user-1`

Instance ID: i-0705fa39969d0d176

asaadmahmood commented:
We are discussing this in UX.

esethna commented:
Thanks for the PR @DSchalla. 

In terms of the functionality, I'm not certain on how it's supposed to work for channel admins? Given that a channel admin role only lives within a channel. For example, what should be the expected if I open the user pop-over on a username in search results? Since search results are team wide so they may display results from the same user in different channels where the user has different channel roles (Channel admin in Channel A, but normal member in channel B). 

![image](https://user-images.githubusercontent.com/14333569/43978624-26fec5ca-9ca5-11e8-8866-064f77d17117.png)

Also adding @wiersgallak since this is related to work her team is doing.

______

In terms of UI, discussed the design team and propose a small tweak that the icon be changed to a label describing the admin level "System Admin" "Team Admin" or "Channel Admin". @asaadmahmood can help with the styling if you can add the appropriate text.

![image](https://user-images.githubusercontent.com/14333569/43978707-705b5922-9ca5-11e8-8fe4-2d949c3afd22.png)

DSchalla commented:
Considering that channel admins are a EE concept I couldn't test or plan for that limitation accordingly. From search results - in theory - it should use the channel the message was posted from, but that might not be that easily doable - Do we get a "context" from the popup creation event? Happy to hear any proposals.

Re styling: Yeah, makes sense. Looks even better, even when it might be tricky with users that have even longer user handles.

mattermod commented:
Spinmint test running for more than 7 days. This test server was terminated.

mattermod commented:
Spinmint test server destroyed

mattermod commented:
`Setup Test Server` label detected. Spinmint test server created if build succeeds (checks pass and no conflicts with base branch).

mattermod commented:
Spinmint test server created at: https://i-026137979a3d6d5b4.test.spinmint.com

Test Admin Account: Username: `sysadmin` | Password: `sysadmin`

Test User Account: Username: `user-1` | Password: `user-1`

Instance ID: i-026137979a3d6d5b4

esethna commented:
Given that a channel admin only lives within a channel, I think it might be best to only show the channel admin badge on a profile pop-over in the center pane of a channel or in the right-hand side reply thread in a channel. 

Cool, Asaad can help with the styling if you can add the appropriate text for the admin level when you get a chance. Thanks!

mattermod commented:
Spinmint test server destroyed

DSchalla commented:
Sounds good to me @esethna. For the team and channel admins to function some further changes are required after having a chat with @jwilander, we'll need to pass the channel and team reference to the popover function since the team or channel roles are not stored in the user roles, I missed that in the PR version.

Should we aim for a first release in 5.3 with the System Admin label in your proposed styling and then release a second version in 5.4 with team + channel admin? Background for that is that I won't be able to get the redux changes implemented until the change window closes for 5.3. Personally, I'd prefer to have System Admin in 5.3 behind an experimental flag, but that's related to my specific MM instance needs and might be biased.

esethna commented:
Might be better to wait and release the feature in v5.4 so we don't need to rush to make the feature window and have time to implement and test the feature in full form. If that works for you?

DSchalla commented:
Yup, works for me!

On Tue, Aug 21, 2018, 23:00 Eric Sethna <notifications@github.com> wrote:

> Might be better to wait and release the feature in v5.4 so we don't need
> to rush to make the feature window and have time to implement and test the
> feature in full form. If that works for you?
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/mattermost/mattermost-webapp/pull/1487#issuecomment-414820078>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AFshaVVw-dNpuZiQflSUjyiFtSh3wGg8ks5uTHTpgaJpZM4Vku6j>
> .
>

esethna commented:
@DSchalla please let us know when you've had a chance to address the feedback above and we can pull in Asaad for the styling and retest. Thanks!

DSchalla commented:
@esethna Just as an update, not forgotten, I should have some time this and next week to get back to it.

DSchalla commented:
@esethna I've updated the PR, it now supports team admins properly. Channel Admin and updated unit tests are still pending.

@asaadmahmood I've tried to apply the styling as good as I've could based on the screenshot, if any changes are required happy to do so of course.

esethna commented:
Thanks @DSchalla, looks like a build error if you can please have a look?

DSchalla commented:
@esethna Fixed with the help of Harrison. Only pending is now the channel admin addition, I'll try to get that done asap, although the styling and general function could be confirmed to be fine before.

mattermod commented:
`Setup Test Server` label detected. Spinmint test server created if build succeeds (checks pass and no conflicts with base branch).

DSchalla commented:
Sure, sounds good. I'll review the Search Functionality as well, not sure why it shouldn't be working there.

esethna commented:
Hey @DSchalla, let me know if you have any questions on the above ^

mattermod commented:
Spinmint test running for more than 7 days. This test server was terminated.

DSchalla commented:
@esethna I cannot reproduce your issue with the search results locally:

![image](https://user-images.githubusercontent.com/5972329/48168971-bcf57a00-e2f1-11e8-96b2-899d18c0d3c9.png)

![image](https://user-images.githubusercontent.com/5972329/48168953-ab13d700-e2f1-11e8-8f4c-3752f0fb518e.png)

DSchalla commented:
@esethna Channel admin support added - works also in RHS + searches for me, but I could only simulate it and not test it with a real EE yet.

mattermod commented:
Spinmint test server created at: https://i-0577b512124627470.test.spinmint.com

Test Admin Account: Username: `sysadmin` | Password: `sysadmin`

Test User Account: Username: `user-1` | Password: `user-1`

Instance ID: i-0577b512124627470

DSchalla commented:
Changes applied, channel admin badge removed for the search bar.

mattermod commented:
Spinmint test server created at: https://i-049dcc62b52d7b514.test.spinmint.com

Test Admin Account: Username: `sysadmin` | Password: `sysadmin`

Test User Account: Username: `user-1` | Password: `user-1`

Instance ID: i-049dcc62b52d7b514

DSchalla commented:
@esethna Cannot reproduce, this is from the `user-1` user with Chrome (Incognito mode):
![image](https://user-images.githubusercontent.com/5972329/48475674-95c11000-e7fd-11e8-9294-fbd751cece98.png)
![image](https://user-images.githubusercontent.com/5972329/48475736-ba1cec80-e7fd-11e8-9f75-6244d1d4656a.png)

Using Firefox, with `sysadmin`:
![image](https://user-images.githubusercontent.com/5972329/48475794-dde03280-e7fd-11e8-9de1-431aef9441a1.png)

In both browsers working just fine. Regarding the other issue you've mentioned, that's unrelated to my PR itself, I'd suggest we open another issue for that?

DSchalla commented:
Quick update, @lieut-data helped me to find the mistake in the code causing the behavior, although it doesn't explain why it indeed worked for me. Would you be so kind and give it again a shot @esethna?

mattermod commented:
`Setup Test Server` label detected. Spinmint test server created if build succeeds (checks pass and no conflicts with base branch).

mattermod commented:
Spinmint test server created at: https://i-0489794e630ae98eb.test.spinmint.com

Test Admin Account: Username: `sysadmin` | Password: `sysadmin`

Test User Account: Username: `user-1` | Password: `user-1`

Instance ID: i-0489794e630ae98eb

mattermod commented:
Spinmint test server destroyed

mattermod commented:
`Setup Test Server` label detected. Spinmint test server created if build succeeds (checks pass and no conflicts with base branch).

mattermod commented:
Spinmint test server created at: https://i-0cfa0d65ea35a7bc4.test.spinmint.com

Test Admin Account: Username: `sysadmin` | Password: `sysadmin`

Test User Account: Username: `user-1` | Password: `user-1`

Instance ID: i-0cfa0d65ea35a7bc4

DSchalla commented:
Related to the fixes previously introduced for the username cutoff, should be fixed by enforcing a min-height for the username element.

mattermod commented:
`Setup Test Server` label detected. Spinmint test server created if build succeeds (checks pass and no conflicts with base branch).

mattermod commented:
Spinmint test server created at: https://i-0ad75b5f6d6d78b09.test.spinmint.com

Test Admin Account: Username: `sysadmin` | Password: `sysadmin`

Test User Account: Username: `user-1` | Password: `user-1`

Instance ID: i-0ad75b5f6d6d78b09

DSchalla commented:
The difference in the behavior is caused by the `master` screenshot using the Mattermost Desktop app, and the Spinmint using your browser. At least that's for me the case. If I add the Spinmint to my MM Desktop app, the underline is continuous. If I open `master` in Chrome, it's not, like in Spinmint.

mattermod commented:
Spinmint test server destroyed

mattermod commented:
Spinmint test running for more than 7 days. This test server was terminated.

mattermod commented:
Spinmint test running for more than 7 days. This test server was terminated.

grundleborg commented on components/profile_popover/profile_popover.jsx:
```diff
@@ -345,10 +372,23 @@ class ProfilePopover extends React.Component {
 
+        const supportsChannelAdmin = this.props.isLicensed;
+        const isChannelAdmin = supportsChannelAdmin && (Utils.isChannelAdmin(this.props.isLicensed, this.props.channelMember.roles) || this.props.channelMember.scheme_admin);
+
```
I think this check should just be `this.props.channelMember.scheme_admin`. The concept of `channel_admin` as a stand-alone role no longer exists since the new permissions system landed.

grundleborg replied:
Also, I'm not convinced about doing a license check here. Channel Admin's *do* exist in TE, it's just their powers are not configurable.

DSchalla replied:
Makes sense, I've taken the example from https://github.com/mattermost/mattermost-webapp/blob/master/components/channel_members_dropdown/channel_members_dropdown.jsx#L78, should we adapt it there as well as part of a separate PR?

grundleborg replied:
Maybe leave it there for now. It's there to keep backwards compatability while the advanced permissions migration is running. We can delete it in future, but probably not yet. However, it's not needed here as this is a completely new feature, which can remain invisible until the migration completes if someone upgrades from 5.1 directly to 5.6 without any inbetween versions.

grundleborg commented on components/profile_popover/profile_popover.jsx:
```diff
@@ -345,10 +372,23 @@ class ProfilePopover extends React.Component {
+        if (Utils.isSystemAdmin(this.props.user.roles)) {
+            roleTitle = <span className='user-popover__role'>{'System Admin'}</span>;
+        } else if (Utils.isInRole(this.props.teamMember.roles, 'team_admin')) {
+            roleTitle = <span className='user-popover__role'>{'Team Admin'}</span>;
```
This should just be `this.props.teamMember.scheme_admin` as with the channel admin case.

grundleborg commented on components/profile_popover/profile_popover.jsx:
```diff
@@ -345,10 +372,23 @@ class ProfilePopover extends React.Component {
+        } else if (Utils.isInRole(this.props.teamMember.roles, 'team_admin')) {
+            roleTitle = <span className='user-popover__role'>{'Team Admin'}</span>;
+        } else if (isChannelAdmin && !this.props.isSearch) {
+            roleTitle = <span className='user-popover__role'>{'Channel Admin'}</span>;
```
I think these strings should be i18n rather than hard coded in english?

DSchalla replied:
Yes, absolutely. Fixed.

DSchalla commented on components/profile_popover/index.js:
```diff
@@ -5,19 +5,43 @@ import {connect} from 'react-redux';
+    const currentChannel = getCurrentChannel(state);
+
+    let channelId;
+    if (selectedPost.exists === false) {
```
Yes, in my opinion that might be easier. It should be available for the caller every time. I'll adapt the code accordingly.

lieut-data commented on actions/views/profile_popover.js:
```diff
@@ -0,0 +1,28 @@
+
+import {getSelectedPost} from 'selectors/rhs';
+
+export function getMembershipForCurrentEntities(userId) {
```
Beautiful!
